src/Ingests/RedisIngest.php

```diff
@@ -107,4 +108,14 @@ protected function connection(): RedisAdapter
             $this->config->get('pulse.ingest.redis.connection')
         ), $this->config);
     }
+
+    /**
+     * The timestamp, in milliseconds, to trim before.
+     */
+    protected function trimBefore(): int
+    {
+        return CarbonImmutable::now()->subMilliseconds(
+            CarbonInterval::fromString($this->config->get('pulse.ingest.redis.max_retention', '7 days'))->totalMilliseconds
+        )->getTimestampMs();
+    }
```

Comment:
I'm continuing to reach for `CarbonImmutable` here because `CarbonInterval` doesn't have any testing helpers. There is a string format hook, but it is deprecated and feels dicey.